fdbserver/RestoreApplier.actor.cpp

```diff
@@ -151,85 +153,204 @@ ACTOR Future<Void> applyToDB(Reference<RestoreApplierData> self, Database cx) {
+					state MutationRef m;
+					ASSERT_WE_THINK(curIndexInCurTxn < curItInCurTxn->second.size());
+
+					m = curItInCurTxn->second[curIndexInCurTxn];
 					if (m.type >= MutationRef::Type::SetValue && m.type <= MutationRef::Type::MAX_ATOMIC_OP) {
 						typeStr = typeString[m.type];
-					}
-					else {
+					} else {
 						TraceEvent(SevError, "FastRestore").detail("InvalidMutationType", m.type);
 					}
 
+					//TraceEvent(SevDebug, "FastRestore_Debug").detail("ApplierApplyToDB", self->describeNode()).detail("Version", it->first).detail("Mutation", m.toString());
 					if (m.type == MutationRef::SetValue) {
 						tr->set(m.param1, m.param2);
 					} else if (m.type == MutationRef::ClearRange) {
 						KeyRangeRef mutationRange(m.param1, m.param2);
 						tr->clear(mutationRange);
 					} else if (isAtomicOp((MutationRef::Type)m.type)) {
 						tr->atomicOp(m.param1, m.param2, m.type);
+						numAtomicOps++;
 					} else {
 						TraceEvent(SevError, "FastRestore")
 						    .detail("UnhandledMutationType", m.type)
 						    .detail("TypeName", typeStr);
 					}
-					++count;
+
 					transactionSize += m.expectedSize();
 
-					if (transactionSize >= opConfig.transactionBatchSizeThreshold) { // commit per 1000 mutations
-						wait(tr->commit());
-						tr->reset();
-						tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
-						tr->setOption(FDBTransactionOptions::LOCK_AWARE);
-						prevIt = it;
-						prevIndex = index;
-						transactionSize = 0;
+					if (transactionSize >= opConfig.transactionBatchSizeThreshold) { // commit per 512B
+						break; // Got enough mutation in the txn
+					} else {
+						curIndexInCurTxn++;
+						while (curItInCurTxn != self->kvOps.end() && curIndexInCurTxn >= curItInCurTxn->second.size()) {
+							curIndexInCurTxn = 0;
+							curItInCurTxn++;
+							startNextVersion = true;
+						}
+
+						if (startNextVersion || curItInCurTxn == self->kvOps.end()) {
+							break;
+						}
 					}
 				}
+			} // !lastTxnHasError
 
-				if (transactionSize > 0) { // the commit batch should NOT across versions
-					wait(tr->commit());
-					tr->reset();
-					tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
-					tr->setOption(FDBTransactionOptions::LOCK_AWARE);
-					prevIt = it;
-					prevIndex = index;
-					transactionSize = 0;
-				}
-				index = 0;
-			}
-			// Last transaction
-			if (transactionSize > 0) {
+			// Commit the txn and prepare the starting point for next txn
+			if (!lastTxnHasError && (startNextVersion || transactionSize > 0 || curItInCurTxn == self->kvOps.end())) {
 				wait(tr->commit());
 			}
-			break;
+
+			// Logic for a successful transaction: Update current txn info and uncommitted txn info
+			lastTxnHasError = false;
+			curIndexInCurTxn++;
+			while (curItInCurTxn != self->kvOps.end() && curIndexInCurTxn >= curItInCurTxn->second.size()) {
+				curIndexInCurTxn = 0;
+				curItInCurTxn++;
+			}
+			if (curItInCurTxn == self->kvOps.end()) {
+				break;
+			}
```

Comment:
These lines are repeated in line 280-289. It would be better to refactor to get rid of the repetition. One idea is to break at line 251 if `curItInCurTxn == self->kvOps.end()`.

Reply:
These two code blocks are not the same.
The one in 280 to 289 are accumulating mutations to be committed in one transaction.
These lines are preparing the starting mutation (position) for the next transaction. 

It is more straightforward to leave it this way that wrapping everything together, in terms of readability.